google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/AppEngineDeploymentRuntime.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright (C) 2016 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.intellij.appengine.cloud;
+
+import com.google.api.client.repackaged.javax.annotation.OverridingMethodsMustInvokeSuper;
+import com.google.cloud.tools.intellij.util.GctBundle;
+
+import com.intellij.icons.AllIcons.General;
+import com.intellij.openapi.application.ApplicationManager;
+import com.intellij.openapi.progress.ProgressIndicator;
+import com.intellij.openapi.progress.ProgressManager;
+import com.intellij.openapi.progress.Task;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.ui.Messages;
+import com.intellij.remoteServer.runtime.deployment.DeploymentRuntime;
+import com.intellij.remoteServer.runtime.log.LoggingHandler;
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import javax.swing.SwingUtilities;
+
+/**
+ * Takes care of undeploy and stop of App Engine applications.
+ */
+class AppEngineDeploymentRuntime extends DeploymentRuntime {
+
+  private static final String STOP_CONFIRMATION_URI_OPEN_TAG =
+      "<a href='https://cloud.google.com/appengine/docs/java/console/#versions'>";
+  private static final String STOP_CONFIRMATION_URI_CLOSE_TAG = "</a>";
+
+  private Project project;
+  private AppEngineHelper appEngineHelper;
+  private LoggingHandler loggingHandler;
+  private String service;
+  private String version;
+
+  public AppEngineDeploymentRuntime(
+      @NotNull Project project,
+      @NotNull AppEngineHelper appEngineHelper,
+      @NotNull LoggingHandler loggingHandler,
+      @Nullable String service,
+      @Nullable String version) {
+    this.project = project;
+    this.appEngineHelper = appEngineHelper;
+    this.loggingHandler = loggingHandler;
+    this.service = service;
+    this.version = version;
+  }
+
+  @Override
+  public boolean isUndeploySupported() {
+    return service != null && version != null;
```

Comment:
Yes, for our purposes we want both. I think you had it correct. But see my other comment about the possible bug.